pkg/streamingpromql/query.go

```diff
@@ -595,6 +595,12 @@ func (q *Query) Close() {
 	default:
 		panic(fmt.Sprintf("unknown result value type %T", q.result.Value))
 	}
+
+	if q.engine.pedantic {
+		if q.memoryConsumptionTracker.CurrentEstimatedMemoryConsumptionBytes > 0 {
```

Comment:
We should only do this if the query succeeded (`q.result.Err == nil`) - if something goes wrong during the query, not returning everything to the pool is OK.

Reply:
I think this is guaranteed by `if q.result == nil` above. But also, pedantic is only enabled during testing and I think it's reasonable to be pedantic about errors during tests too.

Reply:
`q.result` will be non-nil if the query was evaluated, even if the query failed.

If the query succeeds, `q.result.Err` will be nil and `q.result.Value` will be set.

If the query fails, `q.result.Err` will be non-nil.

>But also, pedantic is only enabled during testing and I think it's reasonable to be pedantic about errors during tests too.

I don't think we should do this: this will force us to ensure that things are returned to the pool even when errors occur, which may make the logic in places more complex than it needs to be.